bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/bufferedmutator/MirroringBufferedMutator.java

```diff
@@ -273,52 +283,91 @@ public final void close() throws IOException {
         return;
       }
 
-      AccumulatedExceptions exceptions = new AccumulatedExceptions();
+      final AccumulatedExceptions exceptions = new AccumulatedExceptions();
 
       try {
+        // Schedule flush of all buffered data and:
+        // synchronous) wait for primary flush to finish;
+        // concurrent) wait for both flushes to finish.
         flushBufferedMutatorBeforeClosing();
-        this.ongoingFlushesCounter.decrementReferenceCount();
-        this.ongoingFlushesCounter
-            .getOnLastReferenceClosed()
-            .get(
-                this.configuration.mirroringOptions.connectionTerminationTimeoutMillis,
-                TimeUnit.MILLISECONDS);
       } catch (InterruptedException | ExecutionException | TimeoutException e) {
         setInterruptedFlagIfInterruptedException(e);
         exceptions.add(new IOException(e));
       }
+
+      // Close the primary buffered mutator, if is flushed in both cases.
       try {
-        this.mirroringTracer.spanFactory.wrapPrimaryOperation(
-            new CallableThrowingIOException<Void>() {
-              @Override
-              public Void call() throws IOException {
-                MirroringBufferedMutator.this.primaryBufferedMutator.close();
-                return null;
-              }
-            },
-            HBaseOperation.BUFFERED_MUTATOR_CLOSE);
+        closePrimaryBufferedMutator();
       } catch (IOException e) {
         exceptions.add(e);
       }
+
+      // We are freeing the initial reference to current level reference counter.
+      referenceCounter.current.decrementReferenceCount();
```

Comment:
Shouldn't we invert these two operations? Can't it happen that the decrement closes the parent?

Reply:
I doesn't matter because these two operations operate on different reference counters.